### PR TITLE
add vk dynamic rendering and validation layers, fix some validation errors

### PR DESCRIPTION
**Pull request by SakuragiHnmch**

### CODE REVIEW

**Review by FlyAndNotDown — CHANGES_REQUESTED** [review]

File: Engine/Source/RHI-Vulkan/Src/Device.cpp [hunk 1]
```diff
@@ -21,6 +21,9 @@
 namespace RHI::Vulkan {
     const std::vector<const char*> DEVICE_EXTENSIONS = {
         "VK_KHR_swapchain",
+        "VK_KHR_dynamic_rendering",
+        "VK_KHR_depth_stencil_resolve",
```
[the remark this answers is not in the document]
Comment (SakuragiHnmch) [reply Engine/Source/RHI-Vulkan/Src/Device.cpp:25]: 这两个扩展不加vaidation layer会报错

File: Engine/Source/RHI-Vulkan/Include/RHI/Vulkan/Gpu.h [hunk 2]
```diff
@@ -9,23 +9,25 @@
...
+        explicit VKGpu(VKInstance& instance, vk::PhysicalDevice vkPhysicalDevice);
         ~VKGpu() override;
 
         GpuProperty GetProperty() override;
         Device* RequestDevice(const DeviceCreateInfo* createInfo) override;
 
         const vk::PhysicalDevice& GetVkPhysicalDevice() const;
 
-        const vk::Instance& GetVKInstance() const;
+        VKInstance* GetInstance() const;
```
Comment (FlyAndNotDown) [inline Engine/Source/RHI-Vulkan/Include/RHI/Vulkan/Gpu.h:25]: 都用引用就行了

File: Engine/Source/RHI-Vulkan/Include/RHI/Vulkan/Instance.h [hunk 3]
```diff
@@ -31,6 +32,7 @@ namespace RHI::Vulkan {
         void PrepareLayers();
         void CreateDebugMessenger();
         void DestroyDebugMessenger();
+        void PopulateDebugMessengerCreateInfo(vk::DebugUtilsMessengerCreateInfoEXT& createInfo);
```
Comment (FlyAndNotDown) [inline Engine/Source/RHI-Vulkan/Include/RHI/Vulkan/Instance.h:35]: 这个函数是不是可以 static，或者就这么几行干脆就没必要封函数了

File: Engine/Source/RHI-Vulkan/Src/CommandEncoder.cpp [hunk 4]
```diff
@@ -60,7 +62,53 @@ namespace RHI::Vulkan {
...
+                    beginInfo->depthStencilAttachment->stencilClearValue}));
+
+            renderingInfo.setPDepthAttachment(&depthStencilAttachmentInfo)
+                .setPStencilAttachment(&depthStencilAttachmentInfo);
+        }
+
+        cmdHandle = cmd.GetVkCommandBuffer();
+        cmdHandle.beginRenderingKHR(&renderingInfo, device.GetGpu()->GetInstance()->GetVkDispatch());
+
+        auto pipeline = dynamic_cast<VKGraphicsPipeline*>(beginInfo->pipeline);
```
Comment (FlyAndNotDown) [inline Engine/Source/RHI-Vulkan/Src/CommandEncoder.cpp:110]: auto*

File: Engine/Source/RHI-Vulkan/Src/CommandEncoder.cpp [hunk 5]
```diff
@@ -76,13 +124,7 @@ namespace RHI::Vulkan {
...
 
         vk::Buffer indexBuffer = mBufferView->GetBuffer().GetVkBuffer();
-        IndexFormat format = mBufferView->GetIndexFormat();
-        vk::IndexType vkFormat;
-        if (format == IndexFormat::UINT16) {
-            vkFormat = vk::IndexType::eUint16;
-        } else if (format == IndexFormat::UINT32) {
-            vkFormat = vk::IndexType::eUint32;
-        }
+        auto vkFormat = VKEnumCast<IndexFormat ,vk::IndexType>(mBufferView->GetIndexFormat());
```
Comment (FlyAndNotDown) [inline Engine/Source/RHI-Vulkan/Src/CommandEncoder.cpp:127]: 这个逗号格式有问题

File: Engine/Source/RHI-Vulkan/Src/Device.cpp (hunk 1 above)
Comment (FlyAndNotDown) [inline Engine/Source/RHI-Vulkan/Src/Device.cpp:25]: "VK_KHR_depth_stencil_resolve" 和 "VK_KHR_create_renderpass2" 确认下是不是一定需要

File: Engine/Source/RHI-Vulkan/Src/TextureView.cpp [hunk 6]
```diff
@@ -17,17 +17,15 @@ namespace RHI::Vulkan {
...
 
-        vk::ImageAspectFlags result = {};
         for (const auto& rule : rules) {
-            if (aspect & rule.first) {
-                result |= rule.second;
+            if (aspect == rule.first) {
+                return rule.second;
             }
         }
-        return result;
```
Comment (FlyAndNotDown) [inline Engine/Source/RHI-Vulkan/Src/TextureView.cpp:26]: 你这最后少了个默认情况的 return 吧，Clion 应该有报 Warning，看一下

File: Engine/Source/RHI-Vulkan/Src/TextureView.cpp [hunk 7]
```diff
@@ -17,17 +17,15 @@ namespace RHI::Vulkan {
...
     }
 
     VKTextureView::VKTextureView(VKTexture& tex, VKDevice& dev, const TextureViewCreateInfo* createInfo)
-        : TextureView(createInfo), device(dev), vkTexture(tex), vkTextureView(VK_NULL_HANDLE)
+        : TextureView(createInfo), device(dev), vkTexture(tex), vkTextureView(VK_NULL_HANDLE), info(*createInfo)
```
Comment (FlyAndNotDown) [inline Engine/Source/RHI-Vulkan/Src/TextureView.cpp:28]: createInfo 是有可能空指针的，建议不要直接存 CreateInfo，把里面需要的东西拆开来存，构造里 Assert(createInfo != nullptr) 然后挨个拷贝

File: Engine/Source/RHI/Include/RHI/TextureView.h [hunk 8]
```diff
@@ -16,6 +16,7 @@ namespace RHI {
         uint8_t mipLevelNum;
         uint8_t baseArrayLayer;
         uint8_t arrayLayerNum;
+        Extent<3> extent;
```
Comment (FlyAndNotDown) [inline Engine/Source/RHI/Include/RHI/TextureView.h:19]: 删掉，从 Texture 里面拿就行了

File: Engine/Source/RHI/Sample/Triangle.cpp [hunk 9]
```diff
@@ -98,6 +98,7 @@ class TriangleApplication : public Application {
             viewCreateInfo.baseMipLevel = 0;
             viewCreateInfo.mipLevelNum = 1;
             viewCreateInfo.aspect = TextureAspect::COLOR;
+            viewCreateInfo.extent = {width, height, 0};
```
Comment (FlyAndNotDown) [inline Engine/Source/RHI/Sample/Triangle.cpp:101]: 删掉

**Review by FlyAndNotDown — CHANGES_REQUESTED** [review]

File: Engine/Source/RHI-Vulkan/Src/Device.cpp (hunk 1 above)
Comment (SakuragiHnmch) [reply Engine/Source/RHI-Vulkan/Src/Device.cpp:25]: **Validation Error: [ VUID-vkCreateDevice-ppEnabledExtensionNames-01387 ] Object 0: handle = 0x7f953c070000, type = VK_OBJECT_TYPE_INSTANCE; | MessageID = 0x12537a2c | Missing extension required by the device extension VK_KHR_dynamic_rendering: VK_KHR_depth_stencil_resolve. The Vulkan spec states: All required device extensions for each extension in the VkDeviceCreateInfo::ppEnabledExtensionNames list must also be present in that list (https://vulkan.lunarg.com/doc/view/1.3.216.0/mac/1.3-extensions/vkspec.html#VUID-vkCreateDevice-ppEnabledExtensionNames-01387)**

**Validation Error: [ VUID-vkCreateDevice-ppEnabledExtensionNames-01387 ] Object 0: handle = 0x7fae59037800, type = VK_OBJECT_TYPE_INSTANCE; | MessageID = 0x12537a2c | Missing extension required by the device extension VK_KHR_depth_stencil_resolve: VK_KHR_create_renderpass2. The Vulkan spec states: All required device extensions for each extension in the VkDeviceCreateInfo::ppEnabledExtensionNames list must also be present in that list (https://vulkan.lunarg.com/doc/view/1.3.216.0/mac/1.3-extensions/vkspec.html#VUID-vkCreateDevice-ppEnabledExtensionNames-01387)**
dynamic rendering 依赖于 depth_stencil_resolve, depth_resolve_stencil依赖于create_renderpass2

**Review by FlyAndNotDown — APPROVED** [review]

**Pull request closed without merging**